src/librustc_typeck/collect.rs

```diff
@@ -317,11 +319,15 @@ fn type_param_predicates<'a, 'tcx>(
     };
 
     let icx = ItemCtxt::new(tcx, item_def_id);
-    result
-        .predicates
-        .extend(icx.type_parameter_bounds_in_generics(ast_generics, param_id, ty,
-            OnlySelfBounds(true)));
-    result
+    extra_preds.extend(
+        icx.type_parameter_bounds_in_generics(ast_generics, param_id, ty, OnlySelfBounds(true)));
+
+    let mut preds = result.predicates.clone();
+    preds.extend(extra_preds);
```

Comment:
should we maybe check if `extra_preds.is_empty` and avoid the clone?

Reply:
I can't quite tell what the type of `result` is here though

Reply:
The correct solution is `make_mut`, IMO, as discussed on IRC.